[Craftsy] Add support for craftsy site.

Bricio opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

#3140

Added Crafsy site support using BrightcoveNew extractor.

mehq commented:
https://www.craftsy.com/class/sew-your-own-designer-handbag/
this page has seven videos (including preview) but your extractor can extract only one (the preview).

mehq commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,63 @@
+            title = class_preview.get('title')
+            url = f'http://players.brightcove.net/{account_id}/default_default/index.html?videoId={video_id}'
+            entries.append(self.url_result(url, 'BrightcoveNew', video_id, title))
+        else:
```
you don't need an `else` block since all classes should have `lessons`. `preview` is optional though

Bricio commented:
@mehq 
> https://www.craftsy.com/class/sew-your-own-designer-handbag/ this page has seven videos (including preview) but your extractor can extract only one (the preview).

It only have 7 videos for premium users. free users only shows 1 video and 6 links blocked.
![image](https://user-images.githubusercontent.com/216170/160252138-6ed2e31e-95f5-4ba1-9a3a-f28130adf895.png)

Like i said in #3140 
I dont have premium account to see what is the result for premium user. 
If you have it, can you share you account so i can check or can you fix the code?

mehq commented:
@Bricio 
then what about this one: https://www.craftsy.com/class/all-access-estes-park-wool-market/
its free but still downloads only one

Bricio commented:
@mehq 

Changed the code to check if user has access or is free.

Based on your first comment, i assume you have premium account. Can you please check if premium video are working too.

thx

mehq commented:
> Based on your first comment, i assume you have premium account. Can you please check if premium video are working too.

sorry i don't have a premium account. perhaps the person who created the issue can help in this regard

lukemax47 commented:
> @mehq
> 
> Changed the code to check if user has access or is free.
> 
> Based on your first comment, i assume you have premium account. Can you please check if premium video are working too.
> 
> thx

I can share access with you. Message me on Telegram @lukemax or Discord @lukemaxwell#3049.

Bricio commented:
Ok, everything is working.

If cookies are not used, yt-dlp is able to download class preview and free videos.

If cookies are passed, yt-dlp is able to download every video. Including premium.
Login was not implemented because captcha is needed.


---

Full download of 1 premium item.
![image](https://user-images.githubusercontent.com/216170/160747645-eaa706d4-0568-46a7-a00f-e916b3a3b6d6.png)

Tests for 3 playlists ( free without class preview, free with class preview, premium )
https://www.craftsy.com/class/the-midnight-quilt-show-season-5/
https://www.craftsy.com/class/sew-your-own-designer-handbag/
https://www.craftsy.com/class/all-access-estes-park-wool-market/

```
D:\ytdlp\yt-dlp>python -m yt_dlp --test --cookies cookies.txt -a craftsy.txt
[Craftsy] the-midnight-quilt-show-season-5: Downloading webpage
[download] Downloading playlist: The Midnight Quilt Show Season 5
[Craftsy] playlist The Midnight Quilt Show Season 5: Collected 10 videos; downloading 10 of them
[download] Downloading video 1 of 10
[brightcove:new] 6172654501001: Downloading JSON metadata
[brightcove:new] 6172654501001: Downloading m3u8 information
[brightcove:new] 6172654501001: Downloading m3u8 information
[brightcove:new] 6172654501001: Downloading MPD manifest
[brightcove:new] 6172654501001: Downloading MPD manifest
[info] 6172654501001: Downloading 1 format(s): hls-4060-1+dash-7b7ff949-264f-4796-9396-6b989ae07ba5-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 70
[download] Destination: Cupid's Arrow [6172654501001].fhls-4060-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Cupid's Arrow [6172654501001].fdash-7b7ff949-264f-4796-9396-6b989ae07ba5-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Cupid's Arrow [6172654501001].mp4"
Deleting original file Cupid's Arrow [6172654501001].fdash-7b7ff949-264f-4796-9396-6b989ae07ba5-1.m4a (pass -k to keep)
Deleting original file Cupid's Arrow [6172654501001].fhls-4060-1.mp4 (pass -k to keep)
[download] Downloading video 2 of 10
[brightcove:new] 6172652811001: Downloading JSON metadata
[brightcove:new] 6172652811001: Downloading m3u8 information
[brightcove:new] 6172652811001: Downloading m3u8 information
[brightcove:new] 6172652811001: Downloading MPD manifest
[brightcove:new] 6172652811001: Downloading MPD manifest
[info] 6172652811001: Downloading 1 format(s): hls-4056-1+dash-e481f6d4-4d61-45b8-851f-6da10f13b8a2-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 78
[download] Destination: Milky Way [6172652811001].fhls-4056-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Milky Way [6172652811001].fdash-e481f6d4-4d61-45b8-851f-6da10f13b8a2-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Milky Way [6172652811001].mp4"
Deleting original file Milky Way [6172652811001].fdash-e481f6d4-4d61-45b8-851f-6da10f13b8a2-1.m4a (pass -k to keep)
Deleting original file Milky Way [6172652811001].fhls-4056-1.mp4 (pass -k to keep)
[download] Downloading video 3 of 10
[brightcove:new] 6172654503001: Downloading JSON metadata
[brightcove:new] 6172654503001: Downloading m3u8 information
[brightcove:new] 6172654503001: Downloading m3u8 information
[brightcove:new] 6172654503001: Downloading MPD manifest
[brightcove:new] 6172654503001: Downloading MPD manifest
[info] 6172654503001: Downloading 1 format(s): hls-4061-1+dash-fb425208-02a8-412d-a448-266f5eea9963-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 70
[download] Destination: Quilt-As-You-Go Placemats [6172654503001].fhls-4061-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Quilt-As-You-Go Placemats [6172654503001].fdash-fb425208-02a8-412d-a448-266f5eea9963-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Quilt-As-You-Go Placemats [6172654503001].mp4"
Deleting original file Quilt-As-You-Go Placemats [6172654503001].fhls-4061-1.mp4 (pass -k to keep)
Deleting original file Quilt-As-You-Go Placemats [6172654503001].fdash-fb425208-02a8-412d-a448-266f5eea9963-1.m4a (pass -k to keep)
[download] Downloading video 4 of 10
[brightcove:new] 6172652812001: Downloading JSON metadata
[brightcove:new] 6172652812001: Downloading m3u8 information
[brightcove:new] 6172652812001: Downloading m3u8 information
[brightcove:new] 6172652812001: Downloading MPD manifest
[brightcove:new] 6172652812001: Downloading MPD manifest
[info] 6172652812001: Downloading 1 format(s): hls-4057-1+dash-42f25c91-9193-4d5d-a9e3-bef08d2bafeb-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 78
[download] Destination: Weave it Be [6172652812001].fhls-4057-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Weave it Be [6172652812001].fdash-42f25c91-9193-4d5d-a9e3-bef08d2bafeb-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Weave it Be [6172652812001].mp4"
Deleting original file Weave it Be [6172652812001].fdash-42f25c91-9193-4d5d-a9e3-bef08d2bafeb-1.m4a (pass -k to keep)
Deleting original file Weave it Be [6172652812001].fhls-4057-1.mp4 (pass -k to keep)
[download] Downloading video 5 of 10
[brightcove:new] 6172652503001: Downloading JSON metadata
[brightcove:new] 6172652503001: Downloading m3u8 information
[brightcove:new] 6172652503001: Downloading m3u8 information
[brightcove:new] 6172652503001: Downloading MPD manifest
[brightcove:new] 6172652503001: Downloading MPD manifest
[info] 6172652503001: Downloading 1 format(s): hls-4053-1+dash-04534c0d-21c1-4691-b101-3e978ab29902-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 81
[download] Destination: Split Square [6172652503001].fhls-4053-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Split Square [6172652503001].fdash-04534c0d-21c1-4691-b101-3e978ab29902-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Split Square [6172652503001].mp4"
Deleting original file Split Square [6172652503001].fdash-04534c0d-21c1-4691-b101-3e978ab29902-1.m4a (pass -k to keep)
Deleting original file Split Square [6172652503001].fhls-4053-1.mp4 (pass -k to keep)
[download] Downloading video 6 of 10
[brightcove:new] 6172653750001: Downloading JSON metadata
[brightcove:new] 6172653750001: Downloading m3u8 information
[brightcove:new] 6172653750001: Downloading m3u8 information
[brightcove:new] 6172653750001: Downloading MPD manifest
[brightcove:new] 6172653750001: Downloading MPD manifest
[info] 6172653750001: Downloading 1 format(s): hls-4043-1+dash-8db0fe2f-572c-4fd3-b6ea-925efe6284a8-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 58
[download] Destination: Bear Tracks [6172653750001].fhls-4043-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Bear Tracks [6172653750001].fdash-8db0fe2f-572c-4fd3-b6ea-925efe6284a8-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Bear Tracks [6172653750001].mp4"
Deleting original file Bear Tracks [6172653750001].fhls-4043-1.mp4 (pass -k to keep)
Deleting original file Bear Tracks [6172653750001].fdash-8db0fe2f-572c-4fd3-b6ea-925efe6284a8-1.m4a (pass -k to keep)
[download] Downloading video 7 of 10
[brightcove:new] 6172654504001: Downloading JSON metadata
[brightcove:new] 6172654504001: Downloading m3u8 information
[brightcove:new] 6172654504001: Downloading m3u8 information
[brightcove:new] 6172654504001: Downloading MPD manifest
[brightcove:new] 6172654504001: Downloading MPD manifest
[info] 6172654504001: Downloading 1 format(s): hls-4061-1+dash-d364a888-b32d-4544-8455-2d4729f48656-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 80
[download] Destination: Twinkling Twilight [6172654504001].fhls-4061-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Twinkling Twilight [6172654504001].fdash-d364a888-b32d-4544-8455-2d4729f48656-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Twinkling Twilight [6172654504001].mp4"
Deleting original file Twinkling Twilight [6172654504001].fdash-d364a888-b32d-4544-8455-2d4729f48656-1.m4a (pass -k to keep)
Deleting original file Twinkling Twilight [6172654504001].fhls-4061-1.mp4 (pass -k to keep)
[download] Downloading video 8 of 10
[brightcove:new] 6172652813001: Downloading JSON metadata
[brightcove:new] 6172652813001: Downloading m3u8 information
[brightcove:new] 6172652813001: Downloading m3u8 information
[brightcove:new] 6172652813001: Downloading MPD manifest
[brightcove:new] 6172652813001: Downloading MPD manifest
[info] 6172652813001: Downloading 1 format(s): hls-4061-1+dash-df6c50db-2707-4cd8-969f-2579223645eb-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 99
[download] Destination: Modern X [6172652813001].fhls-4061-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Modern X [6172652813001].fdash-df6c50db-2707-4cd8-969f-2579223645eb-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Modern X [6172652813001].mp4"
Deleting original file Modern X [6172652813001].fdash-df6c50db-2707-4cd8-969f-2579223645eb-1.m4a (pass -k to keep)
Deleting original file Modern X [6172652813001].fhls-4061-1.mp4 (pass -k to keep)
[download] Downloading video 9 of 10
[brightcove:new] 6172653751001: Downloading JSON metadata
[brightcove:new] 6172653751001: Downloading m3u8 information
[brightcove:new] 6172653751001: Downloading m3u8 information
[brightcove:new] 6172653751001: Downloading MPD manifest
[brightcove:new] 6172653751001: Downloading MPD manifest
[info] 6172653751001: Downloading 1 format(s): hls-4051-1+dash-b6b54e40-77ad-4cf7-9d48-d358ca845e62-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 31
[download] Destination: Tips for Using Your Walking Foot [6172653751001].fhls-4051-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Tips for Using Your Walking Foot [6172653751001].fdash-b6b54e40-77ad-4cf7-9d48-d358ca845e62-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Tips for Using Your Walking Foot [6172653751001].mp4"
Deleting original file Tips for Using Your Walking Foot [6172653751001].fhls-4051-1.mp4 (pass -k to keep)
Deleting original file Tips for Using Your Walking Foot [6172653751001].fdash-b6b54e40-77ad-4cf7-9d48-d358ca845e62-1.m4a (pass -k to keep)
[download] Downloading video 10 of 10
[brightcove:new] 6172652815001: Downloading JSON metadata
[brightcove:new] 6172652815001: Downloading m3u8 information
[brightcove:new] 6172652815001: Downloading m3u8 information
[brightcove:new] 6172652815001: Downloading MPD manifest
[brightcove:new] 6172652815001: Downloading MPD manifest
[info] 6172652815001: Downloading 1 format(s): hls-4059-1+dash-7bafcfe8-c602-40cd-8ed0-da576eec798b-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 34
[download] Destination: Stash Busting Pieced Quilt Back [6172652815001].fhls-4059-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Stash Busting Pieced Quilt Back [6172652815001].fdash-7bafcfe8-c602-40cd-8ed0-da576eec798b-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Stash Busting Pieced Quilt Back [6172652815001].mp4"
Deleting original file Stash Busting Pieced Quilt Back [6172652815001].fhls-4059-1.mp4 (pass -k to keep)
Deleting original file Stash Busting Pieced Quilt Back [6172652815001].fdash-7bafcfe8-c602-40cd-8ed0-da576eec798b-1.m4a (pass -k to keep)
[download] Finished downloading playlist: The Midnight Quilt Show Season 5
[Craftsy] sew-your-own-designer-handbag: Downloading webpage
[download] Downloading playlist: Sew Your Own Designer Handbag
[Craftsy] playlist Sew Your Own Designer Handbag: Collected 7 videos; downloading 7 of them
[download] Downloading video 1 of 7
[brightcove:new] 6286979065001: Downloading JSON metadata
[brightcove:new] 6286979065001: Downloading m3u8 information
[brightcove:new] 6286979065001: Downloading m3u8 information
[brightcove:new] 6286979065001: Downloading MPD manifest
[brightcove:new] 6286979065001: Downloading MPD manifest
[info] 6286979065001: Downloading 1 format(s): hls-4030-1+dash-67f29a88-c890-404e-9257-0080173a745b-1
[download] Sew Your Own Designer Handbag [6286979065001].mp4 has already been downloaded
[download] Downloading video 2 of 7
[brightcove:new] 6286978687001: Downloading JSON metadata
[brightcove:new] 6286978687001: Downloading m3u8 information
[brightcove:new] 6286978687001: Downloading m3u8 information
[brightcove:new] 6286978687001: Downloading MPD manifest
[brightcove:new] 6286978687001: Downloading MPD manifest
[info] 6286978687001: Downloading 1 format(s): hls-4045-1+dash-373f4567-22fa-426c-8a4c-6ae533fc57bb-1
[download] Overview & Tools Needed [6286978687001].mp4 has already been downloaded
[download] Downloading video 3 of 7
[brightcove:new] 6286982272001: Downloading JSON metadata
[brightcove:new] 6286982272001: Downloading m3u8 information
[brightcove:new] 6286982272001: Downloading m3u8 information
[brightcove:new] 6286982272001: Downloading MPD manifest
[brightcove:new] 6286982272001: Downloading MPD manifest
[info] 6286982272001: Downloading 1 format(s): hls-4059-1+dash-fb39e3b9-db67-46ff-9b95-1935fcdbfb70-1
[download] Basic Purse Construction [6286982272001].mp4 has already been downloaded
[download] Downloading video 4 of 7
[brightcove:new] 6286979066001: Downloading JSON metadata
[brightcove:new] 6286979066001: Downloading m3u8 information
[brightcove:new] 6286979066001: Downloading m3u8 information
[brightcove:new] 6286979066001: Downloading MPD manifest
[brightcove:new] 6286979066001: Downloading MPD manifest
[info] 6286979066001: Downloading 1 format(s): hls-4060-1+dash-d4672e60-fe40-4f2d-9386-39a4897a472d-1
[download] Design Option #1 [6286979066001].mp4 has already been downloaded
[download] Downloading video 5 of 7
[brightcove:new] 6286977258001: Downloading JSON metadata
[brightcove:new] 6286977258001: Downloading m3u8 information
[brightcove:new] 6286977258001: Downloading m3u8 information
[brightcove:new] 6286977258001: Downloading MPD manifest
[brightcove:new] 6286977258001: Downloading MPD manifest
[info] 6286977258001: Downloading 1 format(s): hls-4060-1+dash-eaa2fc02-2c23-4f2d-a0b1-928fbf7a2edb-1
[download] Design Option #2 [6286977258001].mp4 has already been downloaded
[download] Downloading video 6 of 7
[brightcove:new] 6286981964001: Downloading JSON metadata
[brightcove:new] 6286981964001: Downloading m3u8 information
[brightcove:new] 6286981964001: Downloading m3u8 information
[brightcove:new] 6286981964001: Downloading MPD manifest
[brightcove:new] 6286981964001: Downloading MPD manifest
[info] 6286981964001: Downloading 1 format(s): hls-4043-1+dash-440d0358-2cfc-45f3-abb9-1369fb8452aa-1
[download] Design Option #3 [6286981964001].mp4 has already been downloaded
[download] Downloading video 7 of 7
[brightcove:new] 6286978792001: Downloading JSON metadata
[brightcove:new] 6286978792001: Downloading m3u8 information
[brightcove:new] 6286978792001: Downloading m3u8 information
[brightcove:new] 6286978792001: Downloading MPD manifest
[brightcove:new] 6286978792001: Downloading MPD manifest
[info] 6286978792001: Downloading 1 format(s): hls-4060-1+dash-3e67542f-4502-40f3-9059-b72b9105d4f3-1
[download] Meet Your Instructor [6286978792001].mp4 has already been downloaded
[download] Finished downloading playlist: Sew Your Own Designer Handbag
[Craftsy] all-access-estes-park-wool-market: Downloading webpage
[download] Downloading playlist: All Access: Estes Park Wool Market
[Craftsy] playlist All Access: Estes Park Wool Market: Collected 6 videos; downloading 6 of them
[download] Downloading video 1 of 6
[brightcove:new] 6172661052001: Downloading JSON metadata
[brightcove:new] 6172661052001: Downloading m3u8 information
[brightcove:new] 6172661052001: Downloading m3u8 information
[brightcove:new] 6172661052001: Downloading MPD manifest
[brightcove:new] 6172661052001: Downloading MPD manifest
[info] 6172661052001: Downloading 1 format(s): hls-4042-1+dash-14a0f351-8563-4284-8ac6-d576d85ba857-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 4
[download] Destination: Bluprint All Access - Estes Park Wool Market [6172661052001].fhls-4042-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Bluprint All Access - Estes Park Wool Market [6172661052001].fdash-14a0f351-8563-4284-8ac6-d576d85ba857-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Bluprint All Access - Estes Park Wool Market [6172661052001].mp4"
Deleting original file Bluprint All Access - Estes Park Wool Market [6172661052001].fdash-14a0f351-8563-4284-8ac6-d576d85ba857-1.m4a (pass -k to keep)
Deleting original file Bluprint All Access - Estes Park Wool Market [6172661052001].fhls-4042-1.mp4 (pass -k to keep)
[download] Downloading video 2 of 6
[brightcove:new] 6172661053001: Downloading JSON metadata
[brightcove:new] 6172661053001: Downloading m3u8 information
[brightcove:new] 6172661053001: Downloading m3u8 information
[brightcove:new] 6172661053001: Downloading MPD manifest
[brightcove:new] 6172661053001: Downloading MPD manifest
[info] 6172661053001: Downloading 1 format(s): hls-4046-1+dash-19672eed-d5a0-418b-845d-12c2cf7bff39-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 44
[download] Destination: Ivy Sipes [6172661053001].fhls-4046-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Ivy Sipes [6172661053001].fdash-19672eed-d5a0-418b-845d-12c2cf7bff39-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Ivy Sipes [6172661053001].mp4"
Deleting original file Ivy Sipes [6172661053001].fdash-19672eed-d5a0-418b-845d-12c2cf7bff39-1.m4a (pass -k to keep)
Deleting original file Ivy Sipes [6172661053001].fhls-4046-1.mp4 (pass -k to keep)
[download] Downloading video 3 of 6
[brightcove:new] 6172664492001: Downloading JSON metadata
[brightcove:new] 6172664492001: Downloading m3u8 information
[brightcove:new] 6172664492001: Downloading m3u8 information
[brightcove:new] 6172664492001: Downloading MPD manifest
[brightcove:new] 6172664492001: Downloading MPD manifest
[info] 6172664492001: Downloading 1 format(s): hls-4059-1+dash-116e7ff0-c94c-4679-973f-fe4be29258aa-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 62
[download] Destination: Monique Mullis [6172664492001].fhls-4059-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Monique Mullis [6172664492001].fdash-116e7ff0-c94c-4679-973f-fe4be29258aa-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Monique Mullis [6172664492001].mp4"
Deleting original file Monique Mullis [6172664492001].fdash-116e7ff0-c94c-4679-973f-fe4be29258aa-1.m4a (pass -k to keep)
Deleting original file Monique Mullis [6172664492001].fhls-4059-1.mp4 (pass -k to keep)
[download] Downloading video 4 of 6
[brightcove:new] 6172661054001: Downloading JSON metadata
[brightcove:new] 6172661054001: Downloading m3u8 information
[brightcove:new] 6172661054001: Downloading m3u8 information
[brightcove:new] 6172661054001: Downloading MPD manifest
[brightcove:new] 6172661054001: Downloading MPD manifest
[info] 6172661054001: Downloading 1 format(s): hls-4061-1+dash-16cd1518-98bf-4eaa-a355-38688479282d-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 47
[download] Destination: TeDi Jansen [6172661054001].fhls-4061-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: TeDi Jansen [6172661054001].fdash-16cd1518-98bf-4eaa-a355-38688479282d-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "TeDi Jansen [6172661054001].mp4"
Deleting original file TeDi Jansen [6172661054001].fdash-16cd1518-98bf-4eaa-a355-38688479282d-1.m4a (pass -k to keep)
Deleting original file TeDi Jansen [6172661054001].fhls-4061-1.mp4 (pass -k to keep)
[download] Downloading video 5 of 6
[brightcove:new] 6172661749001: Downloading JSON metadata
[brightcove:new] 6172661749001: Downloading m3u8 information
[brightcove:new] 6172661749001: Downloading m3u8 information
[brightcove:new] 6172661749001: Downloading MPD manifest
[brightcove:new] 6172661749001: Downloading MPD manifest
[info] 6172661749001: Downloading 1 format(s): hls-4053-1+dash-ed5db828-78eb-4f4a-a699-71604e035699-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 60
[download] Destination: Deb Tewell [6172661749001].fhls-4053-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Deb Tewell [6172661749001].fdash-ed5db828-78eb-4f4a-a699-71604e035699-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Deb Tewell [6172661749001].mp4"
Deleting original file Deb Tewell [6172661749001].fdash-ed5db828-78eb-4f4a-a699-71604e035699-1.m4a (pass -k to keep)
Deleting original file Deb Tewell [6172661749001].fhls-4053-1.mp4 (pass -k to keep)
[download] Downloading video 6 of 6
[brightcove:new] 6172661750001: Downloading JSON metadata
[brightcove:new] 6172661750001: Downloading m3u8 information
[brightcove:new] 6172661750001: Downloading m3u8 information
[brightcove:new] 6172661750001: Downloading MPD manifest
[brightcove:new] 6172661750001: Downloading MPD manifest
[info] 6172661750001: Downloading 1 format(s): hls-4057-1+dash-6816e085-6061-4e62-baec-81d9603a2760-1
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 44
[download] Destination: Shatzie Wardall [6172661750001].fhls-4057-1.mp4
[download] 100% of 10.00KiB in 00:00
[dashsegments] Total fragments: 1
[download] Destination: Shatzie Wardall [6172661750001].fdash-6816e085-6061-4e62-baec-81d9603a2760-1.m4a
[download] 100% of 619.00B in 00:00
[Merger] Merging formats into "Shatzie Wardall [6172661750001].mp4"
Deleting original file Shatzie Wardall [6172661750001].fhls-4057-1.mp4 (pass -k to keep)
Deleting original file Shatzie Wardall [6172661750001].fdash-6816e085-6061-4e62-baec-81d9603a2760-1.m4a (pass -k to keep)
[download] Finished downloading playlist: All Access: Estes Park Wool Market

D:\ytdlp\yt-dlp>
```

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,73 @@
+        webpage = self._download_webpage(url, video_name)
+
+        extra_data = get_element_by_id('vidstore-classes_class-video-player-js-extra', webpage)
+        video_data = json.loads(js_to_json(self._search_regex(
```
Use `self._parse_json` with a `transform_source`

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,73 @@
+    }]
+
+    def _real_extract(self, url):
+        video_name = self._match_id(url)
```
```suggestion
        video_id = self._match_id(url)
```

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,73 @@
+            url = f'http://players.brightcove.net/{account_id}/default_default/index.html?videoId={video_id}'
+            entries.append(self.url_result(url, 'BrightcoveNew', video_id, title))
+
+        if is_free or user_has_access:
```
```suggestion
        if dict_get(video_data, ('is_free', 'user_has_access')):
```

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,73 @@
+            entries.append(self.url_result(url, 'BrightcoveNew', video_id, title))
+
+        if is_free or user_has_access:
+            lessons = video_data.get('lessons')
+            for lesson in lessons:
+                video_id = lesson.get('video_id')
+                title = lesson.get('title')
+                url = f'http://players.brightcove.net/{account_id}/default_default/index.html?videoId={video_id}'
+                entries.append(self.url_result(url, 'BrightcoveNew', video_id, title))
```
```suggestion
            entries += [
                self.url_result(
                    f'http://players.brightcove.net/{account_id}/default_default/index.html?videoId={lesson["video_id"]}',
                    BrightcoveNewIE, lesson['video_id'], lesson.get('title'))
                for lesson in video_data['lessons']]
```
similar above

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,73 @@
+            extra_data, 'video data')))
+
+        account_id = traverse_obj(video_data, ('video_player', 'bc_account_id'))
+        class_title = video_data.get('class_title')
+        description = self._html_search_meta(['og:description', 'description'], webpage, default=None)
+
+        is_free = video_data.get('is_free')
+        user_has_access = video_data.get('user_has_access')
+        class_preview = traverse_obj(video_data, ('video_player', 'class_preview'))
```
none of these need to be variables (since they are only used once). Just use them directly where needed

Bricio replied:
class_preview is used in 3 places, so I keep it.

https://github.com/yt-dlp/yt-dlp/blob/17872d78c1ade7461d8429067cd00f56c289e325/yt_dlp/extractor/craftsy.py#L54-L60

Just changed the line to be just before IF.

pukkandan commented on yt_dlp/extractor/craftsy.py:
```diff
@@ -0,0 +1,74 @@
+                    BrightcoveNewIE, lesson['video_id'], lesson.get('title'))
+                for lesson in video_data['lessons']]
+
+        return self.playlist_result(
+            entries,
+            video_id,
+            video_data.get('class_title'),
+            self._html_search_meta(['og:description', 'description'], webpage, default=None)
+        )
```
```suggestion
        return self.playlist_result(
            entries, video_id, video_data.get('class_title'),
            self._html_search_meta(('og:description', 'description'), webpage, default=None))
```
